### PR TITLE
Add NPS & Polyphony stats

### PR DESCRIPTION
**Pull request by MyBlackMIDIScore**
~Haven't figured out how to do polyphony for cake yet~

Also I am not 100% sure about the NPS implementation, if there is a better way to do it lmk

### CODE REVIEW

File: src/gui/window/stats.rs [hunk 1]
```diff
@@ -163,8 +171,67 @@ impl GuiWasabiWindow {
...
+                            }
+                        }
                     };
                 }
             });
     }
 }
+
+#[derive(Default)]
+pub struct NpsCounter(VecDeque<(Instant, i64)>);
```
Comment (arduano) [inline src/gui/window/stats.rs:206]: I prefer a field for this even if it's just a single field, more descriptive. Tuple structs are only good for very very primitive wrappers around stuff. This is more complex.

File: src/gui/window/stats.rs [hunk 2]
```diff
@@ -163,8 +171,67 @@ impl GuiWasabiWindow {
...
+            if front.0.elapsed().as_secs_f64() > Self::NPS_WINDOW {
+                self.0.pop_front();
+            } else {
+                break;
+            }
+        }
+    }
+
+    pub fn read(&self) -> u32 {
+        let old = if let Some(front) = self.0.front() {
```
Comment (arduano) [inline src/gui/window/stats.rs:223]:
```suggestion
        let old = if let Some((front_time, passed)) = self.0.front() {
```
better than using tuples directly

File: src/gui/window/scene/note_list_system/mod.rs [hunk 3]
```diff
@@ -120,64 +121,69 @@ impl NoteRenderer {
...
-                                            column.key,
-                                            note.color.as_u32(),
-                                            column.border_width as u32,
-                                        );
-
-                                        if note.start <= 0.0
-                                            && column.color.is_none()
-                                            && note.start + note.len > 0.0
-                                        {
+                    let out_data = columns_view_info.par_iter_mut().rev().map(|column| {
```
Comment (arduano) [inline src/gui/window/scene/note_list_system/mod.rs:126]: this is getting complex enough where I think a struct is better. Define a simple struct somewhere above, use it here to pass the data back.

File: src/gui/window/fps.rs [hunk 4]
```diff
@@ -3,33 +3,25 @@ use std::{
     time::{Duration, Instant},
 };
 
-pub struct Fps {
-    frames: VecDeque<Instant>,
-    current: Instant,
-}
+pub struct Fps(VecDeque<Instant>);
```
Comment (arduano) [inline src/gui/window/fps.rs:6]: Use named field pls

**Review by arduano — CHANGES_REQUESTED** [review]

**Comment by MyBlackMIDIScore** [comment]
Oh oops, I forgot about this too

Btw should we release 1.0 and leave #74 and #75 for later or should we fix those first?

**Comment by arduano** [comment]
Yeah 1.0 now is ok

**Pull request merged**